private/bufpkg/bufconnect/interceptors.go

```diff
@@ -82,25 +71,66 @@ func NewAuthorizationInterceptorProvider(container app.EnvContainer) func(string
-//
-// Note that the interceptor returned from this provider is always applied LAST in the series of interceptors added to
-// a client.
-func NewAuthorizationInterceptorProviderWithToken(token string) func(string) connect.UnaryInterceptorFunc {
-	return func(address string) connect.UnaryInterceptorFunc {
-		interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-			return connect.UnaryFunc(func(
-				ctx context.Context,
-				req connect.AnyRequest,
-			) (connect.AnyResponse, error) {
-				if token != "" {
-					req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
-				}
-				return next(ctx, req)
-			})
+// TokenSet is used to provide authentication token in NewAuthorizationInterceptorProvider
+type TokenSet struct {
+	envKey          string
+	bufToken        string
+	remoteUsernames map[string]string
+	remoteTokens    map[string]string
+}
+
+// NewTokenSetFromContainer creates a TokenSet from the BUF_TOKEN environment variable
+func NewTokenSetFromContainer(container app.EnvContainer) (*TokenSet, error) {
+	bufToken := container.Env(tokenEnvKey)
+	tokenSet, err := NewTokenSetFromString(bufToken)
+	if err != nil {
+		return nil, err
+	}
+	tokenSet.envKey = tokenEnvKey
+	return tokenSet, nil
+}
+
+// NewTokenSetFromString creates a TokenSet by the token provided
+func NewTokenSetFromString(token string) (*TokenSet, error) {
+	tokenSet := &TokenSet{
+		remoteUsernames: make(map[string]string),
+		remoteTokens:    make(map[string]string),
+	}
+	tokens := strings.Split(token, ",")
+	for _, u := range tokens {
+		if contain := strings.ContainsAny(u, "@"); contain {
+			keyPairsAndRemoteAddress := strings.Split(u, "@")
+			if len(keyPairsAndRemoteAddress) != 2 {
+				return nil, fmt.Errorf("cannot parse token: %s, invalid remote token: %s", token, u)
+			}
+			keyPairs := strings.Split(keyPairsAndRemoteAddress[0], ":")
+			if len(keyPairs) != 2 {
+				return nil, fmt.Errorf("cannot parse token: %s, invalid remote token: %s", token, u)
+			}
+			remoteAddress := keyPairsAndRemoteAddress[1]
+			username := keyPairs[0]
+			remoteToken := keyPairs[1]
+			if _, ok := tokenSet.remoteTokens[remoteAddress]; ok {
+				return nil, fmt.Errorf("cannot parse token: %s, repeated token for same BSR remote: %s", remoteToken, remoteAddress)
+			}
+			if _, ok := tokenSet.remoteUsernames[remoteAddress]; ok {
+				return nil, fmt.Errorf("cannot parse token: %s, repeated token for same BSR remote: %s", remoteToken, remoteAddress)
+			}
+			tokenSet.remoteTokens[remoteAddress] = remoteToken
+			tokenSet.remoteUsernames[remoteAddress] = username
+		} else {
+			if tokenSet.bufToken != "" {
+				return nil, fmt.Errorf("cannot parse token: %s, two buf token provided: %q and %q", token, u, tokenSet.bufToken)
+			}
+			tokenSet.bufToken = u
 		}
-		return interceptor
 	}
+	return tokenSet, nil
+}
+
+func (t *TokenSet) obtainTokenFromRemoteAddress(remoteAddress string) (token string) {
```

Comment:
`tokenForRemoteAddress(remoteAddress string) string {`

"obtain" isn't different from saying "get"

"from" implies the value is derived from remoteAddress when it's really a lookup.

You didn't use the named return value `token`. No need to have it either.

Reply:
The function name is changed to `lookupRemoteToken`